e2e/app/page/notebook-cell.ts

```diff
@@ -0,0 +1,152 @@
+    const frame = await this.getIFrame();
+    const cell = await frame.waitForSelector(selector, {visible: true});
+    await cell.click({delay: 10}); // focus
+    return cell;
+  }
+
+  /**
+   * Set focus in last cell.
+   */
+  async focusLastCell(): Promise<ElementHandle> {
+    const selector = `${this.cellSelector()} .CodeMirror-code`;
+    const frame = await this.getIFrame();
+    const cells = await frame.$$(selector);
+    this.cellIndex = cells.length;
+    const cell = cells[cells.length - 1];
+    await cell.click({delay: 10}); // focus
+    return cell;
+  }
+
+  /**
+   * Returns cell run result. Result is either stdout or stderr.
+   * @param {number} timeOut The timeout time in milliseconds.
+   */
+  async getRunResult(timeOut: number = 30 * 1000): Promise<string> {
+    return await Promise.race([
+      this.getRunResultSuccessText(timeOut),
+      this.getRunResultErrorText(timeOut),
+    ]);
+  }
+
+  /**
+   * Gets cell output texts.
+   * @param {number} timeOut The timeout time in milliseconds.
+   */
+  async getRunResultSuccessText(timeOut?: number): Promise<string> {
+    const element = await this.findOutputElement(timeOut);
+    const value = await getPropValue<string>(element, 'innerText');
+    await element.dispose();
+    return value.trim();
+  }
+
+  /**
+   * Gets cell output error.
+   * @param {number} timeOut The timeout time in milliseconds.
+   */
+  async getRunResultErrorText(timeOut?: number): Promise<string> {
+    const element = await this.findOutputErrorElement(timeOut);
+    const value = await getPropValue<string>(element, 'innerText');
+    await element.dispose();
+    console.error(`Run cell output error: \n${value}`);
+    return value.trim();
+  }
+
+  /**
+   * Find cell output_area element.
+   * @param {number} timeOut The timeout time in milliseconds.
+   */
+  async findOutputElement(timeOut?: number): Promise<ElementHandle> {
+    const selector = `${this.outputAreaSelector(this.getCellIndex())}:not(.output_error)`;
+    const iframe = await this.getIFrame();
+    await iframe.waitForSelector(selector, {visible: true, timeout: timeOut});
+    const elements = await iframe.$$(selector);
+    return elements[elements.length - 1];
+  }
+
+  /**
+   * Find cell output_area error.
+   * @param {number} timeOut The timeout time in milliseconds.
+   */
+  async findOutputErrorElement(timeOut?: number): Promise<ElementHandle> {
+    const selector = `${this.outputAreaSelector(this.getCellIndex())}.output_error`;
+    const iframe = await this.getIFrame();
+    return iframe.waitForSelector(selector, {visible: true, timeout: timeOut});
+  }
+
+  getCellIndex(): number {
+    return this.cellIndex;
+  }
+
+  outputAreaSelector(indx?: number): string {
+    return `${this.cellSelector(indx)} .output_subarea`;
+  }
+
+  /**
+   * Returns css selector for cell.
+   * @param {number} indx Cell index.
+   */
+  cellSelector(indx?: number): string {
```

Comment:
I would probably make some of these private for clarity. I doubt there's a need to use this outside this class.